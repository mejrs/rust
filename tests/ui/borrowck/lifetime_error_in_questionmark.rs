//@ compile-flags: -Z track-diagnostics

#![crate_type= "lib"]

use std::error::Error;
use std::fmt::{Debug, Display, Formatter};

#[derive(Debug)]
pub struct MyError<Src> {
    stuff: Src,
}
impl<T: Debug> Display for MyError<T> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
        Debug::fmt(self, f)
    }
}
impl<T: Debug> Error for MyError<T> {}

// like `anyhow::Error`
#[derive(Debug)]
struct StaticError;

impl<E> From<E> for StaticError
where
    E: Error + Send + Sync + 'static,
{
    fn from(_error: E) -> Self {
        StaticError
    }
}

fn parse<X>(stuff: X) -> Result<(), MyError<X>> {
    Err(MyError { stuff })
}

fn foo() -> Result<(), StaticError> {
    let stuff = vec![1_u8, 2, 3, 4];
    parse(&stuff)?;
    //~^ ERROR `stuff` does not live long enough
    Ok(())
}
