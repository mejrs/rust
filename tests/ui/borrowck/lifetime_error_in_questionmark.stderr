[compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:2855:9] &place_desc = Some(
    "stuff",
)
[compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:2855:9] &explanation = MustBeValidFor {
    category: CallArgument(
        Some(
            FnDef(
                DefId(0:16 ~ lifetime_error_in_questionmark[30f4]::parse),
                [
                    &'{erased} std::vec::Vec<u8, std::alloc::Global>,
                ],
            ),
        ),
    ),
    from_closure: false,
    span: $DIR/lifetime_error_in_questionmark.rs:38:5: 38:18 (#0),
    region_name: RegionName {
        name: "'static",
        source: Static,
    },
    opt_place_desc: Some(
        "stuff",
    ),
    path: [
        ('?2: '?9) due to Single(bb4[6]) (None) (Boring),
        ('?9: '?3) due to Single(bb4[7]) (None) (CallArgument(Some(FnDef(DefId(0:16 ~ lifetime_error_in_questionmark[30f4]::parse), [&'{erased} std::vec::Vec<u8, std::alloc::Global>])))),
        ('?3: '?8) due to Single(bb4[7]) (None) (Boring),
        ('?8: '?4) due to Single(bb5[1]) (None) (CallArgument(Some(FnDef(DefId(2:4092 ~ core[d7b7]::ops::try_trait::Try::branch), [std::result::Result<(), MyError<&'{erased} std::vec::Vec<u8, std::alloc::Global>>>])))),
        ('?4: '?7) due to Single(bb5[1]) (None) (Boring),
        ('?7: '?5) due to Single(bb10[1]) (None) (Boring),
        ('?5: '?10) due to Single(bb10[1]) (None) (Assignment),
        ('?10: '?11) due to Single(bb10[3]) (None) (Boring),
        ('?11: '?6) due to Single(bb10[4]) (None) (CallArgument(Some(FnDef(DefId(2:4095 ~ core[d7b7]::ops::try_trait::FromResidual::from_residual), [std::result::Result<(), StaticError>, std::result::Result<std::convert::Infallible, MyError<&'{erased} std::vec::Vec<u8, std::alloc::Global>>>])))),
        ('?6: '?0) due to Single(bb10[4]) (None) (Predicate(no-location (#0))),
    ],
}
error[E0597]: `stuff` does not live long enough
  --> $DIR/lifetime_error_in_questionmark.rs:38:11
   |
LL |     parse(&stuff)?;
   |     ------^^^^^^-- this use of the `?` operator requires that `stuff` can be borrowed for `'static`
...
LL | }
   | - `stuff` dropped here while still borrowed
-Ztrack-diagnostics: created at compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:2977:28

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0597`.
